### ISSUE TITLE
Process.join seem to block until timeout if a Popen has been called before

### ISSUE THREAD

**Issue by Champax** [gevent/gevent#600]
Process.join seem to block until timeout if a Popen has been called before.
(gevent 1.1a1)

If i call only "test_process" (without calling "test_invoke" before), it's fine :

```
2015-07-02 14:38:41,608 | INFO join in
2015-07-02 14:38:41,608 | INFO sleep in
2015-07-02 14:38:42,610 | INFO sleep out
2015-07-02 14:38:42,631 | INFO join out

Process finished with exit code 0
```

If whenever a call to "test_invoke" is fired, followed by "test_process", the "p.join" only exit after timeout is reached.
```
2015-07-02 14:39:17,157 | INFO invoke in
2015-07-02 14:39:17,160 | INFO got out='invoke_ok'
2015-07-02 14:39:17,160 | INFO invoke out
2015-07-02 14:39:17,165 | INFO join in
2015-07-02 14:39:17,165 | INFO sleep in
2015-07-02 14:39:18,165 | INFO sleep out
2015-07-02 14:39:27,166 | INFO join out ===> ### 9 sec delay...

Process finished with exit code 0
```

Code for repro :
```
"""
# -*- coding: utf-8 -*-
"""
from gevent import monkey
monkey.patch_all()

import logging
import sys
logging.basicConfig(level="INFO")
f = logging.Formatter("%(asctime)s | %(levelname)s %(message)s")
c = logging.StreamHandler(sys.stdout)
c.setLevel(logging.getLevelName("INFO"))
c.setFormatter(f)
root = logging.getLogger()
root.setLevel(logging.getLevelName("INFO"))
root.handlers = []
root.addHandler(c)

import gevent
from multiprocessing import Process
from gevent.subprocess import Popen, PIPE

logger = logging.getLogger("Test")


def test_invoke():
    """
    Test
    """

    logger.info("invoke in")
    p = Popen("echo -n 'invoke_ok'".split(' '), stdout=PIPE, stderr=PIPE)
    gevent.sleep(0)
    so, _ = p.communicate()
    logger.info("got out=%s", so)
    gevent.sleep(0)
    logger.info("invoke out")


def f(sleep_sec):
    """
    Test
    :param sleep_sec: sleep_ms
    """
    logger.info("sleep in")
    gevent.sleep(sleep_sec)
    logger.info("sleep out")


def test_process():
    """
    Test
    """

    # Launch
    p = Process(target=f, args=(1.0,))
    p.start()

    logger.info("join in")
    p.join(10)
    logger.info("join out")


if __name__ == '__main__':
    # Enabling this cause test_process to hang in join() until timeout is reached even if process exit before
    test_invoke()

    test_process()
```

**Comment by thedrow** [comment]
Did it happen before with 1.0?

**Comment by Champax** [comment]
Hum not tested this before gevent 1.0.2

**Comment by jamadden** [comment]
1.0.x in general, I think is what thedrow means. The major difference is that `subprocess` is not monkey patched by default under 1.0.x but is under 1.1a1. So the questions are:

- Does it work by default under 1.0.x? If yes, does it still work if you do `patch_all(subprocess=True)`?
- Does it work under 1.1a1 if you do `patch_all(subprocess=False)`?

**Comment by Champax** [comment]
Quick check : same behavior in all cases (will re-do more tests later ony)

gevent 1.0.2

```
2015-07-02 18:25:52,095 | INFO gevent == 1.0.2
2015-07-02 18:25:52,095 | INFO invoke in
2015-07-02 18:25:52,098 | INFO got out='invoke_ok'
2015-07-02 18:25:52,098 | INFO invoke out
2015-07-02 18:25:52,102 | INFO join in
2015-07-02 18:25:52,103 | INFO sleep in
2015-07-02 18:25:53,103 | INFO sleep out
2015-07-02 18:26:02,104 | INFO join out
```

gevent 1.0.2 + subprocess=True

```
2015-07-02 18:28:42,090 | INFO gevent + patch == 1.0.2
2015-07-02 18:28:42,090 | INFO invoke in
2015-07-02 18:28:42,093 | INFO got out='invoke_ok'
2015-07-02 18:28:42,093 | INFO invoke out
2015-07-02 18:28:42,098 | INFO join in
2015-07-02 18:28:42,098 | INFO sleep in
2015-07-02 18:28:43,099 | INFO sleep out
2015-07-02 18:28:52,099 | INFO join out
```

gevent 1.1a1 + subprocess=False

```
2015-07-02 18:30:52,625 | INFO gevent == 1.1a1
2015-07-02 18:30:52,625 | INFO invoke in
2015-07-02 18:30:52,628 | INFO got out='invoke_ok'
2015-07-02 18:30:52,629 | INFO invoke out
2015-07-02 18:30:52,632 | INFO join in
2015-07-02 18:30:52,633 | INFO sleep in
2015-07-02 18:30:53,633 | INFO sleep out
2015-07-02 18:31:02,633 | INFO join out
```

**Comment by jamadden** [comment]
Ok, great, at least it's not a new issue.

**Comment by Champax** [comment]
Please note that calling join without timeout works fine, ie

```
logger.info("join in")
p.join()
logger.info("join out")
```

result in

```
2015-07-07 23:52:43,114 | INFO invoke in
2015-07-07 23:52:43,117 | INFO got out='invoke_ok'
2015-07-07 23:52:43,118 | INFO invoke out
2015-07-07 23:52:43,119 | INFO join in
2015-07-07 23:52:43,120 | INFO sleep in
2015-07-07 23:52:44,121 | INFO sleep out
2015-07-07 23:52:44,121 | INFO join out
```

**Comment by jamadden** [comment]
I am unable to reproduce this locally using your original script. Can you still reproduce this with current master?

```
2015-07-08 07:47:13,109 | INFO invoke in
2015-07-08 07:47:13,119 | INFO got out=b"'invoke_ok'"
2015-07-08 07:47:13,119 | INFO invoke out
2015-07-08 07:47:13,122 | INFO join in
2015-07-08 07:47:13,124 | INFO sleep in
2015-07-08 07:47:14,125 | INFO sleep out
2015-07-08 07:47:14,128 | INFO join out
```

**Comment by thedrow** [comment]
Are you guys on the same operating system? Does it matter?

**Comment by jamadden** [comment]
That's a great question. It was going to be my next one :) My local reproduction test was on OS X 10.10.

**Comment by Champax** [comment]
Still occurs on ubuntu (3.13.0-55-generic #94-Ubuntu SMP Thu Jun 18 00:27:10 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux)
I will retry on debian 8, python 2.7.9 when i have time

```
['gevent==1.1a3.dev0', 'greenlet==0.4.7']
python 2.7.6 (default, Jun 22 2015, 17:58:13) 
[GCC 4.8.2]
gevent 1.1a3.dev0
2015-07-09 11:22:12,420 | INFO invoke in
2015-07-09 11:22:12,424 | INFO got out='invoke_ok'
2015-07-09 11:22:12,424 | INFO invoke out
2015-07-09 11:22:12,426 | INFO join in
2015-07-09 11:22:12,436 | INFO sleep in
2015-07-09 11:22:13,437 | INFO sleep out
2015-07-09 11:22:22,427 | INFO join out
```

**Comment by jamadden** [comment]
Ok, I got a chance to look into this on Linux I was able to reproduce it. It seems it's another variation of the old `os.waitpid` vs `gevent.subprocess` (or indeed, any libev child watcher) thing, as mentioned [here](https://github.com/gevent/gevent/issues/452#issuecomment-119316538) and tangentially [here](https://github.com/gevent/gevent/issues/280#issuecomment-120123286).

Basically, inside `Process.wait`, multiprocessing wants to use `os.waitpid` in a tight loop, polling for the end of process. But if you've already used `gevent.subprocess`, `os.waitpid` becomes useless. Sadly, multiprocessing ignores the error return code from `waitpid` that means "no such process", assuming it means that the process hasn't started yet (not that it has already exited). (I'm not quite sure why I don't see this on OS X, though. Perhaps it's because the Linux version of libev uses signalfd for more reliable signal handling and queueing.)

It's not entirely clear to me how to handle this. I *think* it may be possible to hook up gevent.subprocess with a monkey-patched os.waitpid, but I'm not sure.
